BWEnv/src/controller.cc

```diff
@@ -586,31 +587,36 @@ int Controller::getAttackFrames(int unitID) {
   return attackFrames;
 }
 
-void Controller::serializeFrameData(torchcraft::fbs::FrameDataT* data) {
-  std::ostringstream out;
-  if (prev_sent_frame == nullptr) {
-    out << *last_frame;
-    data->is_diff = false;
-  } else {
-    out << replayer::frame_diff(last_frame, prev_sent_frame);
-    data->is_diff = true;
+void Controller::serializeFrameData(torchcraft::fbs::FrameOrFrameDiffUnion& frameOrFrameDiff) {
+  frameOrFrameDiff.Reset();
+  {    
+    flatbuffers::FlatBufferBuilder builder;
+    if (prev_sent_frame == nullptr) {
+      frameOrFrameDiff.type = torchcraft::fbs::FrameOrFrameDiff::Frame;
+      last_frame->addToFlatBufferBuilder(builder);
+    } else {
+      frameOrFrameDiff.type = torchcraft::fbs::FrameOrFrameDiff::FrameDiff;
+      auto frameDiff = replayer::frame_diff(last_frame, prev_sent_frame);
+      frameDiff.addToFlatBufferBuilder(builder);      
+    }
+    frameOrFrameDiff.value = builder.GetBufferPointer();
   }
-  if (prev_sent_frame != nullptr)
+  
+  if (prev_sent_frame == nullptr) {
```

Comment:
How can you decref a nullptr?

Reply:
Poorly :) I'd switched this to if(prev_sent_frame) then switched it back... incorrectly.

Fortunately, this has the relatively easy-to-detect effect of *breaking everything*